### DISCUSSION TITLE
This is (one way) of getting the base URL for a remote API from Blazor

### DISCUSSION THREAD

**Discussion by sturlath** [abpframework/abp#11044] [Show and tell] [answered]
I needed to explicitly set the URL to an API because I'm using this [upload components](https://blazor.syncfusion.com/documentation/file-upload/getting-started) from Syncfusion that has a savUrl parameter that needs to be set with full url. 

To do this you can inject the following into YourPage.razor.cs page (if you are at least using [Dynamic client proxies](https://docs.abp.io/en/abp/latest/API/Dynamic-CSharp-API-Clients))

```csharp
  private readonly AbpRemoteServiceOptions abpRemoteServiceOptions;
  public YourPage(IOptions<AbpRemoteServiceOptions> abpRemoteServiceOptions)
  {
      this.abpRemoteServiceOptions = abpRemoteServiceOptions.Value;
  }
```
and then in your code you can access the values

```csharp
  var remoteServiceConfig = abpRemoteServiceOptions.RemoteServices.GetConfigurationOrDefault("Default"); //or "BookStore"
  var url = remoteServiceConfig .BaseUrl;
```

Is this the best way? Is it the only way? I don't know! I couldn't find any mention in the Issues,Discussions or in the support forum that could help me.. maybe there is something there but its not easily found..

I got to this conclusion after trying out `IHttpClientFactory ` ,`AbpHttpClientOptions ` , `IServiceProvider ` etc... maybe you think this should be obvious and hey I found it out BUT it took me time I wish I could have spent on something else..

I just wanted to share this code because this is the stuff I find most frustrating with the abp.io documentation. These "corner" cases that I have to spend long time on each time I go out of the beaten path.

Don´t get me wrong... I like apb.io but the documentation for **lots of basic stuff just isn´t that good** (even though there is a lot of things in there)

### How to make the documentation better?
I **suggest** that we could add comments on each documentation page (just like Microsoft docks have) where users can comment and contribute their issues and solutions. You can then decide to augment the docs.

But yes hopefully this will help somebody down the line...

**Comment by EngincanV** [comment, marked as answer]
Hi @sturlath, I think it's the most suitable way. `AbpRemoteServiceOptions` sets automatically but you can also configure it, you can read the [relevant section](https://docs.abp.io/en/abp/latest/API/Dynamic-CSharp-API-Clients#abpremoteserviceoptions) from **Dynamic C# API Client Proxies** documentation.

* If you want to configure it, you can define a remote service name and configure it like below:

```csharp
Configure<AbpRemoteServiceOptions>(options =>
{
    options.RemoteServices["MyRemoteServiceName"] = new RemoteServiceConfiguration
    {
        BaseUrl = "...."
     };
});
```

---

> How to make the documentation better?
I suggest that we could add comments on each documentation page (just like Microsoft docks have) where users can comment and contribute their issues and solutions. You can then decide to augment the docs.

Thanks for your suggestion.

We've added something similar to our document platform, you can give your feedback by using it. 
(It shows up bottom-right of the document.)

* You can rate your experience,

![docs-abp-feedback](https://user-images.githubusercontent.com/43685404/147322276-1085b44a-fe07-4c72-a09a-6139c60a1349.png)

* Also, you can add comments to help us how we can improve the documentation.

![docs-abp-feedback-comment](https://user-images.githubusercontent.com/43685404/147322277-acfe5229-597c-4b96-be88-f544fee0a985.png)

**Reply by sturlath** [reply]
Ok thanks for the confirmation and pointing out the document comments its a great start! 

But wouldn't it be great if we could see what others are battling with? But great start and I hope this feature will be enhanced as time goes.

**Reply by EngincanV** [reply]
I will share your suggestion with the team 👍
Thanks again @sturlath  :pray:

**Reply by sturlath** [reply]
@EngincanV can I somewhoe get base url for  Web.public address up to the Blazor application? 

I need it to be able to open a public-web-page in separate tab to the Blazor one.

I probably could do this (but it seams wasteful) 

1.  Add another remoteservice (not going to use it)
2.  Read the [App:MVCPublicUrl] base url from config
3.  Set BaseUrl as the MVCPublicUrl

**or** create a rest API that reads the config and returns it. Sound like the best way if there isn't another default way I´m missing.

Or is there another better way?

**Reply by maliming** [reply]
> Read the [App:MVCPublicUrl] base url from config

This will be the easiest. anyway this is not related to ABP.

**Reply by sturlath** [reply]
I can't read that config directly from Blazor since it is set in the Host and I don´t want to config it 2x (I could of course and that might be easier). I implemented a API endpoint.

But I thought I'd ask if ABP did have any inbuilt methods of getting config. Like I could have used RemoteServices above and set BaseUrl and used ...RemoteServices.GetConfigurationOrDefault() just to get that base url. So its not a stretch to think you have some method of doing that.

But I´m good, will use an API for now
